File: backend/src/domain/models/therapy.py
"""Phase 3c Models — Therapieplan, Konsilien, Arztbriefe, Pflegediagnosen,
Schichtübergabe, Ernährung, Verbrauchsmaterial."""

import uuid
from datetime import UTC, date, datetime

from sqlalchemy import Boolean, Date, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.infrastructure.database import Base


# ─── Therapieplan ──────────────────────────────────────────────

class TreatmentPlan(Base):
    """Therapieplan — ärztlich verordneter Behandlungsplan."""
    __tablename__ = "treatment_plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    title: Mapped[str] = mapped_column(String(255))
    diagnosis: Mapped[str] = mapped_column(Text)
    icd_code: Mapped[str | None] = mapped_column(String(20))

    goals: Mapped[str] = mapped_column(Text)
    interventions: Mapped[str] = mapped_column(Text)

    start_date: Mapped[date] = mapped_column(Date)
    target_date: Mapped[date | None] = mapped_column(Date)
    end_date: Mapped[date | None] = mapped_column(Date)

    status: Mapped[str] = mapped_column(String(20), default="active")  # active, completed, cancelled, on-hold
    priority: Mapped[str] = mapped_column(String(20), default="normal")  # low, normal, high, urgent

    responsible_physician_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    review_date: Mapped[date | None] = mapped_column(Date)
    notes: Mapped[str | None] = mapped_column(Text)
    extra: Mapped[dict | None] = mapped_column(JSONB)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC),
    )

    items: Mapped[list["TreatmentPlanItem"]] = relationship(
        back_populates="plan", cascade="all, delete-orphan", order_by="TreatmentPlanItem.sort_order",
    )


class TreatmentPlanItem(Base):
    """Einzelne Massnahme innerhalb eines Therapieplans."""
    __tablename__ = "treatment_plan_items"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    plan_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("treatment_plans.id", ondelete="CASCADE"), index=True)

    item_type: Mapped[str] = mapped_column(String(30))  # medication, physiotherapy, lab, imaging, nursing, other
    description: Mapped[str] = mapped_column(Text)
    frequency: Mapped[str | None] = mapped_column(String(100))
    duration: Mapped[str | None] = mapped_column(String(100))

    status: Mapped[str] = mapped_column(String(20), default="pending")  # pending, in_progress, completed, cancelled
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    notes: Mapped[str | None] = mapped_column(Text)

    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

    plan: Mapped["TreatmentPlan"] = relationship(back_populates="items")


# ─── Konsilien ─────────────────────────────────────────────────

class Consultation(Base):
    """Konsil — Anfrage an eine Fachdisziplin."""
    __tablename__ = "consultations"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    specialty: Mapped[str] = mapped_column(String(100))  # kardiologie, neurologie, chirurgie, etc.
    urgency: Mapped[str] = mapped_column(String(20), default="routine")  # routine, urgent, emergency
    question: Mapped[str] = mapped_column(Text)
    clinical_context: Mapped[str | None] = mapped_column(Text)

    requested_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    requested_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

    consultant_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    consultant_name: Mapped[str | None] = mapped_column(String(200))

    response: Mapped[str | None] = mapped_column(Text)
    recommendations: Mapped[str | None] = mapped_column(Text)
    responded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    status: Mapped[str] = mapped_column(String(20), default="requested")  # requested, accepted, completed, cancelled

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC),
    )


# ─── Arztbriefe ────────────────────────────────────────────────

class MedicalLetter(Base):
    """Arztbrief — Entlass-/Zuweisungs-/Verlaufsbriefe."""
    __tablename__ = "medical_letters"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    letter_type: Mapped[str] = mapped_column(String(30))  # discharge, referral, progress, transfer
    title: Mapped[str] = mapped_column(String(255))

    recipient_name: Mapped[str | None] = mapped_column(String(200))
    recipient_institution: Mapped[str | None] = mapped_column(String(255))
    recipient_email: Mapped[str | None] = mapped_column(String(255))

    diagnosis: Mapped[str | None] = mapped_column(Text)
    history: Mapped[str | None] = mapped_column(Text)
    findings: Mapped[str | None] = mapped_column(Text)
    therapy: Mapped[str | None] = mapped_column(Text)
    procedures: Mapped[str | None] = mapped_column(Text)
    recommendations: Mapped[str | None] = mapped_column(Text)
    medication_on_discharge: Mapped[str | None] = mapped_column(Text)
    follow_up: Mapped[str | None] = mapped_column(Text)

    content: Mapped[str | None] = mapped_column(Text)  # Freitext-Fallback

    author_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    co_signed_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    co_signed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    status: Mapped[str] = mapped_column(String(20), default="draft")  # draft, final, sent, amended
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    sent_via: Mapped[str | None] = mapped_column(String(30))  # hin_mail, email, fax, print

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC),
    )


# ─── Pflegediagnosen ──────────────────────────────────────────

class NursingDiagnosis(Base):
    """Pflegediagnose nach NANDA-I Klassifikation."""
    __tablename__ = "nursing_diagnoses"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    nanda_code: Mapped[str | None] = mapped_column(String(20))
    title: Mapped[str] = mapped_column(String(255))
    domain: Mapped[str | None] = mapped_column(String(100))  # NANDA-Domain

    defining_characteristics: Mapped[str | None] = mapped_column(Text)
    related_factors: Mapped[str | None] = mapped_column(Text)
    risk_factors: Mapped[str | None] = mapped_column(Text)

    goals: Mapped[str | None] = mapped_column(Text)  # NOC-Outcomes
    interventions: Mapped[str | None] = mapped_column(Text)  # NIC-Interventionen
    evaluation: Mapped[str | None] = mapped_column(Text)

    priority: Mapped[str] = mapped_column(String(20), default="normal")  # low, normal, high
    status: Mapped[str] = mapped_column(String(20), default="active")  # active, resolved, inactive

    diagnosed_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    diagnosed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC),
    )


# ─── Schichtübergabe ──────────────────────────────────────────

class ShiftHandover(Base):
    """Strukturierte Schichtübergabe zwischen Pflegepersonal."""
    __tablename__ = "shift_handovers"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    shift_type: Mapped[str] = mapped_column(String(20))  # early, late, night
    handover_date: Mapped[date] = mapped_column(Date, index=True)

    # SBAR-Struktur
    situation: Mapped[str] = mapped_column(Text)
    background: Mapped[str | None] = mapped_column(Text)
    assessment: Mapped[str | None] = mapped_column(Text)
    recommendation: Mapped[str | None] = mapped_column(Text)

    open_tasks: Mapped[dict | None] = mapped_column(JSONB)  # [{task, priority, due}]
    critical_info: Mapped[str | None] = mapped_column(Text)

    handed_over_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    received_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))
    acknowledged_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))


# ─── Ernährung ─────────────────────────────────────────────────

class NutritionOrder(Base):
    """Ernährungsplan / Diätverordnung."""
    __tablename__ = "nutrition_orders"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    diet_type: Mapped[str] = mapped_column(String(50))  # normal, light, diabetic, renal, parenteral, enteral, npo
    texture: Mapped[str | None] = mapped_column(String(30))  # normal, pureed, liquid, soft
    supplements: Mapped[str | None] = mapped_column(Text)
    restrictions: Mapped[str | None] = mapped_column(Text)  # Allergien, Unverträglichkeiten
    allergies: Mapped[str | None] = mapped_column(Text)

    caloric_target: Mapped[int | None] = mapped_column(Integer)  # kcal/Tag
    protein_target: Mapped[float | None] = mapped_column(Float)  # g/Tag
    fluid_target: Mapped[int | None] = mapped_column(Integer)  # ml/Tag

    special_instructions: Mapped[str | None] = mapped_column(Text)

    status: Mapped[str] = mapped_column(String(20), default="active")  # active, on-hold, completed, cancelled
    start_date: Mapped[date] = mapped_column(Date)
    end_date: Mapped[date | None] = mapped_column(Date)

    ordered_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC), onupdate=lambda: datetime.now(UTC),
    )


class NutritionScreening(Base):
    """Ernährungs-Screening (NRS-2002, MUST, etc.)."""
    __tablename__ = "nutrition_screenings"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)

    screening_type: Mapped[str] = mapped_column(String(30))  # nrs2002, must, mna, sga
    total_score: Mapped[int] = mapped_column(Integer)
    risk_level: Mapped[str] = mapped_column(String(20))  # low, medium, high
    items: Mapped[dict] = mapped_column(JSONB, default=dict)
    notes: Mapped[str | None] = mapped_column(Text)

    screened_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    screened_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))


# ─── Verbrauchsmaterial ────────────────────────────────────────

class SupplyItem(Base):
    """Verbrauchsmaterial-Katalog."""
    __tablename__ = "supply_items"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255))
    article_number: Mapped[str | None] = mapped_column(String(50), unique=True)
    category: Mapped[str] = mapped_column(String(50))  # wound_care, infusion, catheter, respiratory, other
    unit: Mapped[str] = mapped_column(String(20))  # piece, pack, ml, set
    stock_quantity: Mapped[int] = mapped_column(Integer, default=0)
    min_stock: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))


class SupplyUsage(Base):
    """Dokumentation des Verbrauchsmaterial-Einsatzes pro Patient."""
    __tablename__ = "supply_usages"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("patients.id"), index=True)
    supply_item_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("supply_items.id"), index=True)
    encounter_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("encounters.id"))

    quantity: Mapped[int] = mapped_column(Integer)
    reason: Mapped[str | None] = mapped_column(Text)

    used_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))
    used_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

File: packages/shared-types/src/therapy.ts
/**
 * Shared types für Phase 3c — Therapieplan, Konsilien, Arztbriefe,
 * Pflegediagnosen, Schichtübergabe, Ernährung, Verbrauchsmaterial, Dossier.
 */

// ─── Therapieplan ─────────────────────────────────────────────

export type TreatmentPlanStatus = "active" | "completed" | "suspended" | "cancelled";
export type TreatmentPlanPriority = "low" | "normal" | "high" | "urgent";
export type TreatmentPlanItemType = "medication" | "lab" | "procedure" | "monitoring" | "therapy" | "other";

export interface TreatmentPlanItem {
  id: string;
  treatment_plan_id: string;
  item_type: TreatmentPlanItemType;
  description: string;
  frequency?: string;
  target_value?: string;
  is_completed: boolean;
  completed_at?: string;
  completed_by?: string;
  sort_order: number;
}

export interface TreatmentPlan {
  id: string;
  patient_id: string;
  encounter_id?: string;
  title: string;
  diagnosis: string;
  icd_code?: string;
  goals: string;
  interventions: string;
  start_date: string;
  target_end_date?: string;
  status: TreatmentPlanStatus;
  priority: TreatmentPlanPriority;
  items: TreatmentPlanItem[];
  created_by?: string;
  created_at: string;
  updated_at?: string;
}

export interface TreatmentPlanCreate {
  patient_id: string;
  encounter_id?: string;
  title: string;
  diagnosis: string;
  icd_code?: string;
  goals: string;
  interventions: string;
  start_date: string;
  target_end_date?: string;
  priority?: TreatmentPlanPriority;
  items?: { item_type: TreatmentPlanItemType; description: string; frequency?: string; target_value?: string }[];
}

export interface TreatmentPlanUpdate {
  title?: string;
  diagnosis?: string;
  icd_code?: string;
  goals?: string;
  interventions?: string;
  status?: TreatmentPlanStatus;
  priority?: TreatmentPlanPriority;
  target_end_date?: string;
}

export interface PaginatedTreatmentPlans {
  items: TreatmentPlan[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Konsilien ────────────────────────────────────────────────

export type ConsultationSpecialty =
  | "kardiologie" | "pneumologie" | "neurologie" | "gastroenterologie"
  | "nephrologie" | "endokrinologie" | "diabetologie" | "infektiologie"
  | "haematologie" | "onkologie" | "rheumatologie" | "dermatologie"
  | "radiologie" | "chirurgie" | "urologie" | "gynaekologie"
  | "psychiatrie" | "palliativmedizin" | "other";

export type ConsultationStatus = "requested" | "scheduled" | "completed" | "cancelled";
export type ConsultationUrgency = "routine" | "urgent" | "emergency";

export interface Consultation {
  id: string;
  patient_id: string;
  encounter_id?: string;
  specialty: ConsultationSpecialty;
  urgency: ConsultationUrgency;
  status: ConsultationStatus;
  question: string;
  clinical_context?: string;
  response?: string;
  responded_by?: string;
  responded_at?: string;
  requested_by?: string;
  created_at: string;
  updated_at?: string;
}

export interface ConsultationCreate {
  patient_id: string;
  encounter_id?: string;
  specialty: ConsultationSpecialty;
  urgency?: ConsultationUrgency;
  question: string;
  clinical_context?: string;
}

export interface ConsultationUpdate {
  urgency?: ConsultationUrgency;
  question?: string;
  clinical_context?: string;
}

export interface PaginatedConsultations {
  items: Consultation[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Arztbriefe ───────────────────────────────────────────────

export type MedicalLetterType = "discharge" | "referral" | "progress" | "transfer";
export type MedicalLetterStatus = "draft" | "final" | "sent";

export interface MedicalLetter {
  id: string;
  patient_id: string;
  encounter_id?: string;
  letter_type: MedicalLetterType;
  title: string;
  diagnosis?: string;
  anamnesis?: string;
  findings?: string;
  therapy?: string;
  recommendations?: string;
  status: MedicalLetterStatus;
  author_id?: string;
  co_signed_by?: string;
  co_signed_at?: string;
  sent_at?: string;
  sent_to?: string;
  created_at: string;
  updated_at?: string;
}

export interface MedicalLetterCreate {
  patient_id: string;
  encounter_id?: string;
  letter_type: MedicalLetterType;
  title: string;
  diagnosis?: string;
  anamnesis?: string;
  findings?: string;
  therapy?: string;
  recommendations?: string;
}

export interface MedicalLetterUpdate {
  title?: string;
  diagnosis?: string;
  anamnesis?: string;
  findings?: string;
  therapy?: string;
  recommendations?: string;
}

export interface PaginatedMedicalLetters {
  items: MedicalLetter[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Pflegediagnosen ─────────────────────────────────────────

export type NursingDiagnosisStatus = "active" | "resolved" | "inactive";
export type NursingDiagnosisPriority = "low" | "normal" | "high";

export interface NursingDiagnosis {
  id: string;
  patient_id: string;
  encounter_id?: string;
  nanda_code?: string;
  title: string;
  domain?: string;
  defining_characteristics?: string;
  related_factors?: string;
  goals: string;
  interventions: string;
  evaluation?: string;
  status: NursingDiagnosisStatus;
  priority: NursingDiagnosisPriority;
  resolved_at?: string;
  created_by?: string;
  created_at: string;
  updated_at?: string;
}

export interface NursingDiagnosisCreate {
  patient_id: string;
  encounter_id?: string;
  nanda_code?: string;
  title: string;
  domain?: string;
  defining_characteristics?: string;
  related_factors?: string;
  goals: string;
  interventions: string;
  priority?: NursingDiagnosisPriority;
}

export interface NursingDiagnosisUpdate {
  title?: string;
  goals?: string;
  interventions?: string;
  evaluation?: string;
  status?: NursingDiagnosisStatus;
  priority?: NursingDiagnosisPriority;
}

export interface PaginatedNursingDiagnoses {
  items: NursingDiagnosis[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Schichtübergabe ────────────────────────────────────────

export type ShiftType = "early" | "late" | "night";

export interface ShiftHandover {
  id: string;
  patient_id: string;
  encounter_id?: string;
  shift_type: ShiftType;
  handover_date: string;
  situation: string;
  background?: string;
  assessment?: string;
  recommendation?: string;
  open_tasks?: Record<string, unknown>[];
  handed_over_by?: string;
  received_by?: string;
  acknowledged_at?: string;
  created_at: string;
  updated_at?: string;
}

export interface ShiftHandoverCreate {
  patient_id: string;
  encounter_id?: string;
  shift_type: ShiftType;
  handover_date: string;
  situation: string;
  background?: string;
  assessment?: string;
  recommendation?: string;
  open_tasks?: Record<string, unknown>[];
}

export interface PaginatedShiftHandovers {
  items: ShiftHandover[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Ernährung ──────────────────────────────────────────────

export type DietType = "normal" | "light" | "diabetic" | "renal" | "parenteral" | "enteral" | "npo";
export type ScreeningType = "nrs2002" | "must" | "mna" | "sga";

export interface NutritionOrder {
  id: string;
  patient_id: string;
  encounter_id?: string;
  diet_type: DietType;
  texture?: string;
  caloric_target?: number;
  protein_target?: number;
  fluid_target?: number;
  restrictions?: string;
  supplements?: string;
  start_date: string;
  end_date?: string;
  status: "active" | "completed" | "cancelled";
  ordered_by?: string;
  created_at: string;
  updated_at?: string;
}

export interface NutritionOrderCreate {
  patient_id: string;
  encounter_id?: string;
  diet_type: DietType;
  texture?: string;
  caloric_target?: number;
  protein_target?: number;
  fluid_target?: number;
  restrictions?: string;
  supplements?: string;
  start_date: string;
  end_date?: string;
}

export interface NutritionOrderUpdate {
  diet_type?: DietType;
  texture?: string;
  caloric_target?: number;
  protein_target?: number;
  fluid_target?: number;
  restrictions?: string;
  supplements?: string;
  status?: "active" | "completed" | "cancelled";
  end_date?: string;
}

export interface NutritionScreening {
  id: string;
  patient_id: string;
  encounter_id?: string;
  screening_type: ScreeningType;
  total_score: number;
  risk_level: string;
  items?: Record<string, unknown>;
  notes?: string;
  screened_by?: string;
  created_at: string;
}

export interface NutritionScreeningCreate {
  patient_id: string;
  encounter_id?: string;
  screening_type: ScreeningType;
  total_score: number;
  risk_level: string;
  items?: Record<string, unknown>;
  notes?: string;
}

export interface PaginatedNutritionOrders {
  items: NutritionOrder[];
  total: number;
  page: number;
  per_page: number;
}

export interface PaginatedNutritionScreenings {
  items: NutritionScreening[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Verbrauchsmaterial ──────────────────────────────────────

export type SupplyCategory = "wound_care" | "infusion" | "catheter" | "respiratory" | "other";

export interface SupplyItem {
  id: string;
  name: string;
  article_number?: string;
  category: SupplyCategory;
  unit: string;
  stock_quantity: number;
  min_stock: number;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

export interface SupplyItemCreate {
  name: string;
  article_number?: string;
  category: SupplyCategory;
  unit: string;
  stock_quantity?: number;
  min_stock?: number;
}

export interface SupplyItemUpdate {
  name?: string;
  article_number?: string;
  category?: SupplyCategory;
  unit?: string;
  stock_quantity?: number;
  min_stock?: number;
  is_active?: boolean;
}

export interface SupplyUsage {
  id: string;
  patient_id: string;
  supply_item_id: string;
  encounter_id?: string;
  quantity: number;
  reason?: string;
  used_by?: string;
  created_at: string;
}

export interface SupplyUsageCreate {
  patient_id: string;
  supply_item_id: string;
  encounter_id?: string;
  quantity: number;
  reason?: string;
}

export interface PaginatedSupplyItems {
  items: SupplyItem[];
  total: number;
  page: number;
  per_page: number;
}

export interface PaginatedSupplyUsages {
  items: SupplyUsage[];
  total: number;
  page: number;
  per_page: number;
}

// ─── Dossier ────────────────────────────────────────────────

export interface DossierSummary {
  active_alarms: number;
  active_medications: number;
  active_treatment_plans: number;
  active_nursing_diagnoses: number;
  open_consultations: number;
  draft_letters: number;
}

export interface DossierOverview {
  patient_id: string;
  encounter?: Record<string, unknown>;
  summary: DossierSummary;
  latest_vitals?: Record<string, unknown>;
  latest_handover?: Record<string, unknown>;
  active_nutrition?: Record<string, unknown>;
  recent_notes: Record<string, unknown>[];
  recent_nursing: Record<string, unknown>[];
}

// ─── Labels ─────────────────────────────────────────────────

export const TREATMENT_PLAN_STATUS_LABELS: Record<TreatmentPlanStatus, string> = {
  active: "Aktiv",
  completed: "Abgeschlossen",
  suspended: "Pausiert",
  cancelled: "Abgebrochen",
};

export const TREATMENT_PLAN_PRIORITY_LABELS: Record<TreatmentPlanPriority, string> = {
  low: "Niedrig",
  normal: "Normal",
  high: "Hoch",
  urgent: "Dringend",
};

export const CONSULTATION_SPECIALTY_LABELS: Record<ConsultationSpecialty, string> = {
  kardiologie: "Kardiologie",
  pneumologie: "Pneumologie",
  neurologie: "Neurologie",
  gastroenterologie: "Gastroenterologie",
  nephrologie: "Nephrologie",
  endokrinologie: "Endokrinologie",
  diabetologie: "Diabetologie",
  infektiologie: "Infektiologie",
  haematologie: "Hämatologie",
  onkologie: "Onkologie",
  rheumatologie: "Rheumatologie",
  dermatologie: "Dermatologie",
  radiologie: "Radiologie",
  chirurgie: "Chirurgie",
  urologie: "Urologie",
  gynaekologie: "Gynäkologie",
  psychiatrie: "Psychiatrie",
  palliativmedizin: "Palliativmedizin",
  other: "Andere",
};

export const CONSULTATION_STATUS_LABELS: Record<ConsultationStatus, string> = {
  requested: "Angefragt",
  scheduled: "Geplant",
  completed: "Erledigt",
  cancelled: "Abgesagt",
};

export const LETTER_TYPE_LABELS: Record<MedicalLetterType, string> = {
  discharge: "Austrittsbericht",
  referral: "Überweisungsbericht",
  progress: "Zwischenbericht",
  transfer: "Verlegungsbericht",
};

export const LETTER_STATUS_LABELS: Record<MedicalLetterStatus, string> = {
  draft: "Entwurf",
  final: "Finalisiert",
  sent: "Gesendet",
};

export const DIET_TYPE_LABELS: Record<DietType, string> = {
  normal: "Normalkost",
  light: "Leichte Kost",
  diabetic: "Diabetiker-Diät",
  renal: "Nierendiät",
  parenteral: "Parenteral",
  enteral: "Enteral",
  npo: "Nüchtern (NPO)",
};

export const SHIFT_TYPE_LABELS: Record<ShiftType, string> = {
  early: "Frühschicht",
  late: "Spätschicht",
  night: "Nachtschicht",
};

export const SUPPLY_CATEGORY_LABELS: Record<SupplyCategory, string> = {
  wound_care: "Wundversorgung",
  infusion: "Infusion",
  catheter: "Katheter",
  respiratory: "Atemwege",
  other: "Andere",
};


// ─── Medizinische Diagnosen (ICD-10) ──────────────────────────

export type DiagnosisType = "haupt" | "neben" | "verdacht";
export type DiagnosisStatus = "active" | "resolved" | "ruled_out" | "recurrence";
export type DiagnosisSeverity = "leicht" | "mittel" | "schwer";
export type DiagnosisLaterality = "links" | "rechts" | "beidseits";

export interface Diagnosis {
  id: string;
  patient_id: string;
  encounter_id?: string;
  icd_code?: string;
  title: string;
  description?: string;
  diagnosis_type: DiagnosisType;
  severity?: DiagnosisSeverity;
  body_site?: string;
  laterality?: DiagnosisLaterality;
  status: DiagnosisStatus;
  onset_date?: string;
  resolved_date?: string;
  diagnosed_by?: string;
  diagnosed_at: string;
  notes?: string;
  created_at: string;
  updated_at?: string;
}

export interface DiagnosisCreate {
  patient_id: string;
  encounter_id?: string;
  icd_code?: string;
  title: string;
  description?: string;
  diagnosis_type?: DiagnosisType;
  severity?: DiagnosisSeverity;
  body_site?: string;
  laterality?: DiagnosisLaterality;
  onset_date?: string;
  notes?: string;
}

export interface DiagnosisUpdate {
  icd_code?: string;
  title?: string;
  description?: string;
  diagnosis_type?: DiagnosisType;
  severity?: DiagnosisSeverity;
  body_site?: string;
  laterality?: DiagnosisLaterality;
  status?: DiagnosisStatus;
  onset_date?: string;
  resolved_date?: string;
  notes?: string;
}

export interface PaginatedDiagnoses {
  items: Diagnosis[];
  total: number;
  page: number;
  per_page: number;
}

export const DIAGNOSIS_TYPE_LABELS: Record<DiagnosisType, string> = {
  haupt: "Hauptdiagnose",
  neben: "Nebendiagnose",
  verdacht: "Verdachtsdiagnose",
};

export const DIAGNOSIS_STATUS_LABELS: Record<DiagnosisStatus, string> = {
  active: "Aktiv",
  resolved: "Behoben",
  ruled_out: "Ausgeschlossen",
  recurrence: "Rezidiv",
};

export const DIAGNOSIS_SEVERITY_LABELS: Record<DiagnosisSeverity, string> = {
  leicht: "Leicht",
  mittel: "Mittel",
  schwer: "Schwer",
};
